### PR TITLE
disable String(int) ctor

### PR DESCRIPTION
**Pull request by alalek**
**Contrib**: https://github.com/Itseez/opencv_contrib/pull/387

MSVC and GCC compilers interprets cv::String a(0) as a valid
statement with conversion of "int" argument to "const char*".
This patch forbids this expected behaviour.

### CODE REVIEW

File: modules/core/include/opencv2/core/cvstd.hpp [hunk 1]
```diff
@@ -495,10 +495,10 @@ class CV_EXPORTS String
     explicit String();
     String(const String& str);
     String(const String& str, size_t pos, size_t len = npos);
-    String(const char* s);
-    String(const char* s, size_t n);
-    String(size_t n, char c);
-    String(const char* first, const char* last);
+    /* explicit */ String(const char* s);
```
Comment (paroj) [inline modules/core/include/opencv2/core/cvstd.hpp]: why did you add the explicit comment? the `const char*` converting constructor should not be explicit and the two argument constructors can not be explicit.

**Comment by vpisarev** [comment]
@alalek, hmm, would it be more useful to have "explicit String(int)" ?

**Comment by alalek** [comment]
What do you mean exactly? What is expected behaviour in this case?

**Comment by vpisarev** [comment]
sorry for delay. I mean, explicit "String(5)" would be produce string "5". At the same time, the presence of the explicit constructor will make quiet conversion from 0 to String impossible.

**Comment by alalek** [comment]
Actual problem is with this implicit ctor: `String(const char* s);`, but there is no easy way to put here `explicit` modifier (too much code is be broken in this case).
Anyway additional ctor `explicit String(int v);` doesn't help with this problem. Implicit conversion from `int` to `const char*` by compiler is still here for `CommandLineParser::has(0)` call (which accepts only `String` parameter).
This patch forbids compilation of this. Without patch compilation passes, but it is not a valid request.

**Comment by vpisarev** [comment]
got it :+1:
